spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala

```diff
@@ -0,0 +1,161 @@
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn.keras
+
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
+import com.intel.analytics.bigdl.nn.{CAddTable, CAveTable, CMaxTable, CMulTable, CosineDistance, DotProduct, JoinTable, ParallelTable, Sequential => TSequential}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{MultiShape, Shape}
+
+import scala.reflect.ClassTag
+
+class Merge[T: ClassTag](
+   val layers: Array[AbstractModule[Activity, Activity, T]] = null,
+   val mode: String = "sum",
+   val concatAxis: Int = -1,
+   // MultiShape isn't directly supported for serialization. Use Shape instead.
+   val inputShape: Shape = null)(implicit ev: TensorNumeric[T])
+  extends KerasLayer[Tensor[T], Tensor[T], T](Merge.calcBatchInputShape(inputShape, layers)) {
+
+  private val mergeMode = mode.toLowerCase()
+  private var axis = concatAxis
+
+  require(mergeMode == "sum" || mergeMode == "mul" || mergeMode == "concat" || mergeMode == "ave"
+  || mergeMode == "cos" || mergeMode == "dot" || mergeMode == "max",
+  s"Invalid merge mode: $mergeMode")
+  require(layers.length >= 2, s"Merge must have at least two input layers " +
+    s"but found ${layers.length}")
+
+  private def computeOutputShapeForConcat(input: List[Shape], input1: Array[Int]): Shape = {
+    import scala.util.control.Breaks._
+    val output = input1.clone()
+    require(Math.abs(concatAxis) < output.length, s"Invalid concat axis $concatAxis")
+    axis = if (concatAxis < 0) concatAxis + output.length else concatAxis
+    var i = 1
+    while (i < input.length) {
+      val input_i = input(i).toSingle().toArray
+      var j = 0
+      while (j < input_i.length) {
+        if (j != axis) require(input_i(j)==output(j), s"Incompatible input dimension for merge " +
+          s"mode concat: (${output.deep.mkString(", ")}), " +
+          s"(${input_i.deep.mkString(", ")})")
+        j += 1
+      }
+      if (output(axis) == -1 || input_i(axis) == -1) {
+        output(i) = -1
+        break
+      }
+      output(axis) = output(axis) + input_i(axis)
+      i += 1
+    }
+    Shape(output)
+  }
+
+  private def validateInputDim(input: List[Shape], input1: Array[Int]): Unit = {
+    var i = 1
+    while (i < input.length) {
+      val input_i = input(i).toSingle().toArray
+      require(input_i.sameElements(input1), s"Incompatible input dimension for " +
+        s"merge mode $mergeMode: (${input1.deep.mkString(", ")}), " +
+        s"(${input_i.deep.mkString(", ")})")
+      i += 1
+    }
+  }
+
+  override def computeOutputShape(inputShape: Shape): Shape = {
+    val input = inputShape.toMulti()
+    val input1 = input.head.toSingle().toArray
+    if (mergeMode == "concat") {
+      computeOutputShapeForConcat(input, input1)
+    }
+    else {
+      validateInputDim(input, input1)
```

Comment:
we can move `validateInputDim` outside of `if else` block as we need that anyway.